internal/server.go

```diff
@@ -84,6 +101,161 @@ func (s *Server) certManager() *autocert.Manager {
 	}
 }
 
+func (s *Server) localTLSConfig() *tls.Config {
+	return &tls.Config{
+		GetCertificate: s.getLocalCertificate,
+		NextProtos: []string{
+			"h2", "http/1.1", // enable HTTP/2
+		},
+	}
+}
+
+func (s *Server) getLocalCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	name := hello.ServerName
+	if name == "" {
+		return nil, errors.New("thruster_local_tls: missing server name")
+	}
+
+	name, err := idna.Lookup.ToASCII(name)
+	if err != nil {
+		return nil, errors.New("thruster/local_tls: server name contains invalid character")
+	}
+
+	keyUsage := x509.KeyUsageDigitalSignature
+	keyUsage |= x509.KeyUsageKeyEncipherment
+
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		return nil, err
+	}
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization: []string{"Thruster Local"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(365 * 10 * 24 * time.Hour),
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	if ip := net.ParseIP(name); ip != nil {
+		template.IPAddresses = append(template.IPAddresses, ip)
+	} else {
+		template.DNSNames = append(template.DNSNames, name)
+	}
+
+	authority, err := s.getLocalAuthority()
+	if err != nil {
+		return nil, err
+	}
+
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	authcert, err := x509.ParseCertificate(authority.Certificate[0])
+	if err != nil {
+		return nil, err
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, authcert, &priv.PublicKey, authority.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cert := &tls.Certificate{
+		Certificate: [][]byte{authority.Certificate[0], derBytes},
+		PrivateKey:  authority.PrivateKey,
+	}
+
+	return cert, nil
+}
+
+func (s *Server) getLocalAuthority() (*tls.Certificate, error) {
```

Comment:
As mentioned above, I think it's worth introducing a common interface for the two types of TLS provider. At which point, we can move the work of `getLocalAuthority`/`getLocalCertificate` into its own type that implements that interface.

That would also give us a handy place to add some test coverage for `GetCertificate`.